### ISSUE TITLE
Data: Prepare a 20B tokens subsample of Dolma 1.7

### ISSUE THREAD

**Issue by bclavie** [AnswerDotAI/bert24#14]
We've converged on Dolma 1.7 as the baseline dataset for the initial experimentation.

The current versions available are:
- [Full dataset](https://huggingface.co/datasets/allenai/dolma), ~2T token total, with the recommended proportions being to use 1.7T
- [305B subsample by Nous Research](NousResearch/dolma-v1_7-305B), which is useful, but likely too much for all the initial small runs. As we're seemingly going for roughly 2T tokens (+/- future filtering, & multi epochs), it'd be really useful to have a subsample around 20B tokens, i.e. a 1% sample, for experiments.

**Comment by orionw** [comment]
I'm interested in working on this one (and happy to collaborate with others who are interested).  

I assume we should try to sample the 1% to be similar proportions (of web data, code, arxiv, etc.) to the full dataset we intend to use, to be able to ablate data mixtures?

**Comment by bclavie** [comment]
> I assume we should try to sample the 1% to be similar proportions (of web data, code, arxiv, etc.) to the full dataset we intend to use, to be able to ablate data mixtures?

That's exactly it yes! I'd definitely be interesting in helping you out with this too.

> Also @bclavie I can't seem to assign myself to these, perhaps I don't have permissions?
GitHub says you haven't accepted the colaborator invite yet -- can you maybe check your spams? I can resend if it doesn't seem to be anywhere.

**Comment by orionw** [comment]
Yes, my bad - thanks! I can assign now :)

> That's exactly it yes! I'd definitely be interesting in helping you out with this too.

Awesome, this will be a lot of fun!

**Comment by warner-benjamin** [comment]
@bclavie @orionw Nous Research recently uploaded a 30B Dolma 1.7 subset: [NousResearch/dolma-v1_7-30B](https://huggingface.co/datasets/NousResearch/dolma-v1_7-30B)

**Comment by griff4692** [comment]
Is there any literature on the importance of data quality / mixtures for encoder only models?  I only have read the quality filtering for LLM pre-training.

Could imagine smarter masking with something like (https://github.com/microsoft/rho)

**Comment by orionw** [comment]
> Is there any literature on the importance of data quality / mixtures for encoder only models? I only have read the quality filtering for LLM pre-training.

I think most of the prominent data-filtering papers came out after encoders were out of fashion.  I would assume they will work just as well though, but we could run an ablation on that.

Here's an image of the training data comparison for some of the encoders. Note they don't really use web crawl data, except for XLNet.

<img width="951" alt="image" src="https://github.com/AnswerDotAI/bert24/assets/31665361/095361d4-e7aa-4082-89f6-6fe4a8c66fd6">

The XLNet paper says they used "heuristics" to filter, but no further details other than that one sentence. Github issues on it have [been left unanswered](https://github.com/zihangdai/xlnet/issues/212). 

-----------

I was also thinking about data-filtering for the Dolma v1.7 dataset we plan to use.  They group the data into three buckets depending on the quality of the data: head, middle, and tail.  Given how much data there is, I assume we don't want a random sample but would rather sample from head (the highest quality)?  

On the other hand, maybe we should run ablations on that, perhaps noisier docs (but that still passed the basic heuristics) are helpful as well.  What are other's thoughts/intuitions on this?

**Comment by bclavie** [comment]
> Here's an image of the training data comparison for some of the encoders. Note they don't really use web crawl data, except for XLNet.

Empirically we at least know that (good) web crawl data isn't negatively impactful to the performance of the canonical BERT model, as Mosaic was able to fairly straightforwardly reproduce its performance using C4 instead of Wiki+BookC.

> I was also thinking about data-filtering for the Dolma v1.7 dataset we plan to use. They group the data into three buckets depending on the quality of the data: head, middle, and tail. Given how much data there is, I assume we don't want a random sample but would rather sample from head (the highest quality)?
On the other hand, maybe we should run ablations on that, perhaps noisier docs (but that still passed the basic heuristics) are helpful as well. What are other's thoughts/intuitions on this?

Intuitively I'd be tempted to go for the highest quality (since if we do 30% masking, we can run 3 epochs without ever re-masking the same token, we'd still have plenty of tokens to train on), but noisier data is extremely under-explored for encoders, so for all we know, it could help. I'd be in favour of running small-scale experiments on the base size before making a final call here.

**Comment by griff4692** [comment]
> > Here's an image of the training data comparison for some of the encoders. Note they don't really use web crawl data, except for XLNet.
> 
> Empirically we at least know that (good) web crawl data isn't negatively impactful to the performance of the canonical BERT model, as Mosaic was able to fairly straightforwardly reproduce its performance using C4 instead of Wiki+BookC.
> 
> > I was also thinking about data-filtering for the Dolma v1.7 dataset we plan to use. They group the data into three buckets depending on the quality of the data: head, middle, and tail. Given how much data there is, I assume we don't want a random sample but would rather sample from head (the highest quality)?
> > On the other hand, maybe we should run ablations on that, perhaps noisier docs (but that still passed the basic heuristics) are helpful as well. What are other's thoughts/intuitions on this?
> 
> Intuitively I'd be tempted to go for the highest quality (since if we do 30% masking, we can run 3 epochs without ever re-masking the same token, we'd still have plenty of tokens to train on), but noisier data is extremely under-explored for encoders, so for all we know, it could help. I'd be in favour of running small-scale experiments on the base size before making a final call here.

I think small-scale ablations are helpful and we can do them on the smallest scale BERT.  to keep it simple, I would do 4 ablations: High only versus 33/33/33 mix.  If the mix is better we can maybe explore more but first just to test the High quality > diverse hypothesis

**Comment by bclavie** [comment]
> I think small-scale ablations are helpful and we can do them on the smallest scale BERT. to keep it simple, I would do 4 ablations: High only versus 33/33/33 mix. If the mix is better we can maybe explore more but first just to test the High quality > diverse hypothesis

💯 agreed
